vissl/losses/barlow_twins_loss.py

```diff
@@ -0,0 +1,211 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+
+import pprint
+
+import torch
+from classy_vision.generic.distributed_util import all_reduce_mean, get_world_size
+from classy_vision.losses import ClassyLoss, register_loss
+from torch import nn
+from torch.autograd.function import Function
+from vissl.utils.distributed_gradients import gather_from_all
+from vissl.utils.hydra_config import AttrDict
+
+
+class SyncNormalizeFunction(Function):
+    """
+    Adapted from: https://github.com/NVIDIA/apex/blob/master/apex/parallel/sync_batchnorm.py
```

Comment:
curious question: why can't we use the function directly from NVIDIA? if we want to not introduce dependency on apex and that's the reason, that's valid and in that case we could introduce `vissl/utils/apex_helpers.py` and put this function there. This will allow usage beyond this loss function. :)

Reply:
For me there are several advantages to this:
- I think it is important that the criterion is not dependant on `train()` and `eval()` cycles. Using a BatchNorm layer would add hidden running statistics that could lead to weird behaviour. It is possible to disable said statistics using `track_running_stats =False`, however this argument is bugged in the Apex implementation of SyncBatchNorm. See my issue here: [https://github.com/NVIDIA/apex/issues/1071](https://github.com/NVIDIA/apex/issues/1071).
- We do not need to add additional logic to switch between PyTorch SyncBatchNorm and the faster Apex SyncBatchNorm depending on if Apex was installed by the user.

Reply:
> we could introduce vissl/utils/apex_helpers.py and put this function there. This will allow usage beyond this loss function. :)

100 % agree. I wasn't sure where to add it. It also needs to be tested.